guide/src/trait_bounds.md

```diff
@@ -0,0 +1,509 @@
+
+```block
+File "main.py", line 15, in <module>
+   my_rust_model.set_variables_in_rust(2)
+TypeError
+```
+
+It is a type error and Python points to it, so it's easy to identify and solve.
+
+### Wrong types in Python method signatures
+
+Let's assume now that the return type of one of the methods of our Model class is wrong, for example the `get_results` method that is expected to return a `Vec<f64>` in Rust, a list in Python.
+
+```python
+class Model:
+    def set_variables(self, inputs):
+        self.inputs = inputs
+    def compute(self):
+        self.results = [elt**2 -3 for elt in self.inputs]
+    def get_results(self):
+        return self.results[0]
+        #return self.results <-- this is the expected output
+```
+
+This call results in the following panic:
+
+```block
+pyo3_runtime.PanicException: called `Result::unwrap()` on an `Err` value: PyErr { type: Py(0x10dcf79f0, PhantomData) }
+```
+
+This error code is not helpful for a Python user that does not know anything about Rust, or someone that does not know PyO3 was used to interface the Rust code.
+
+However, as we are responsible for making the Rust code available to Python, we can do something about it.
+
+The issue is that we called `unwrap` anywhere we could, and therefore any panic from PyO3 will be directly forwarded to the end user.
+
+Let's modify the code performing the type conversion to give a helpful error message to the Python user:
+
+We used in our `get_results` method the following call that performs the type conversion:
+
+```rust
+impl Model for UserModel {
+  fn get_results(&self) -> Vec<f64> {
+    println!("Get results from Rust calling Python");
+    let gil = Python::acquire_gil();
+    let py = gil.python();
+    self
+        .model
+        .as_ref(py)
+        .call_method("get_results", (), None)
+        .unwrap()
+        .extract()
+        .unwrap()
+}
+```
+
+Let's break it down in order to perform better error handling:
+
+```rust
+impl Model for UserModel {
+  fn get_results(&self) -> Vec<f64> {
+      println!("Get results from Rust calling Python");
+      let gil = Python::acquire_gil();
+      let py = gil.python();
+      let py_result: &PyAny = self
+          .model
+          .as_ref(py)
+          .call_method("get_results", (), None)
+          .unwrap();
+
+      if py_result.get_type().name() != "list" {
+          panic!("Expected a list for the get_results() method signature, got {}", py_result.get_type().name());
+      }
+      py_result.extract().unwrap()
+  }
+}
+```
+
+By doing so, you catch the result of the Python computation and check its type in order to be able to deliver a better error message before performing the unwrapping.
+
+Of course, it does not cover all the possible wrong outputs:
+the user could return a list of strings instead of a list of floats.
+In this case, a runtime panic would still occur due to PyO3, but with an error message much more difficult to decipher for non-rust user.
```

Comment:
Note that we might improve the panic error messages built into pyo3 in the future too 😄

Reply:
👍